### PR TITLE
WIP: Update CellLocator to use latest slicer

### PR DESCRIPTION
**Pull request by allemangD**
Work in progress towards #96.

---

Changes in slicer are in branch [cell-locator-v4.11.0-2020-07-10-b13a49465e](https://github.com/KitwareMedical/Slicer/tree/cell-locator-v4.11.0-2020-07-10-b13a49465e). Similarly with branch [cell-locator-2020-07-10-f0bb5d78](https://github.com/KitwareMedical/CTK/tree/cell-locator-2020-07-10-f0bb5d78) of CTK.

### CODE REVIEW

**Comment by allemangD** [comment]
A few things I've noticed so far that still need to be addressed:

### General

- Annotations have no visual thickness.
- The new [Markups](https://slicer.readthedocs.io/en/latest/user_guide/modules/markups.html) module is meant to support intersection with slices, but such intersections don't always appear.
- Spline color is hard to see against structural pink/red annotations. It should be straightforward to change the markup color, but so far I've not found the correct properties to change.
- Splines/polyline toggle broken. This should be easy to fix with the new markups module, as serialization of splines and polylines are the same, but still needs done.

### CLI Arguments

- combining `--reference-view` with `--view-angle` sometimes causes the 3D view to be empty.
    - For example: `--reference-view Axial --view-angle 15`
- `--annotation-file` still needs updated.

### Keybinds

- `Ctrl+W` (reset), `Ctrl+N` (new) crash the program

### Mouse shortcuts

- Insert annotation (Ctrl+LMB) does not work.
- Translate annotation (LMB drag edge) does not work.
- Scale annotation (RMB drag edge) does not work.

**Comment by jcfr** [comment]
> Annotations have no visual thickness.

The thickness property associated with each annotation should be used to create a model.

To create the model, the following code should be re-used:
* call [vtkSlicerSplinesLogic::CreateModelFromContour](https://github.com/BICCN/cell-locator/blob/6938f2b77e939fbfb4104852cbb2255be57bbf6f/Modules/Loadable/Splines/Logic/vtkSlicerSplinesLogic.cxx#L151-L224) to create a polydata
* associate  the polydata with a model node like it was done in [vtkMRMLMarkupsSplinesDisplayableManager3D::vtkInternal
::UpdateWidgetFromNode](https://github.com/BICCN/cell-locator/blob/6938f2b77e939fbfb4104852cbb2255be57bbf6f/Modules/Loadable/Splines/MRMLDM/vtkMRMLMarkupsSplinesDisplayableManager3D.cxx#L495-L515)

Each time an annotation is added to the scene (NodeAddedEvent), the model should be initialized and observation should be added to call the code referenced above when the annotation is updated. 

Since we the code already exists in C++, I suggest to partially re-introduce  the Splines loadable module and only keep part  of widget and logic. The MRML/MRMLDM/Reader part can be dropped as it has  been re-implemented

> Explore vs Annotate model

image

When annotation mode was toggled from `explore` to `annotate`, the annotation:
* [ ] may be edited by adding/inserting/deleting points. This does not work. Point can still be edited/deleted/... Annotation should be locked.
* [ ] "follows" the slice. This partially work ... after the user add a new point to  the annotation, "follow" does not work. We need to revisit the logic implemented [here](https://github.com/allemangd-forks/cell-locator/blob/34a8e41379901fb2c1286f18bdcb7b42146aca43/Modules/Scripted/Home/Home.py#L1065-L1080)
* [ ] was snapped back to where it was. This does not work.

When `annotate` mode is `enabled`, moving the cursor out of the 3D and back ... expect the user to drop a point. This makes sense because the interaction  mode is set to persistent (`SwitchToPersistentPlaceMode()`) which means that the user is expected to add an other point.

File: Modules/Loadable/Splines/Logic/vtkSlicerSplinesLogic.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  /// Create a 3D slab from the given contour, normal and thickness.
+  static vtkSmartPointer<vtkPolyData> CreateModelFromContour(
+    vtkPolyData* inputContour, vtkVector3d normal, double thickness);
+
+  void BuildSplineModel(vtkMRMLModelNode* modelNode, vtkPolyData *contour, vtkVector3d normal, double thickness);
+
+protected:
+  vtkSlicerSplinesLogic();
+  virtual ~vtkSlicerSplinesLogic();
+
```
Comment (jcfr) [inline Modules/Loadable/Splines/Logic/vtkSlicerSplinesLogic.h]: Add a comment indicating these method are required to satisfy the API logic contract

File: Modules/Loadable/Splines/Logic/vtkSlicerSplinesLogic.cxx [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    );
+
+    modelNode->SetAndObserveMesh(mesh);
+
+    return;
+}
+
+
+//////-----------------------------------------------------------------------------
+//void vtkSlicerSplinesLogic::UpdateSlabModelNode(vtkMRMLMarkupsSplinesNode* splinesNode)
```
Comment (jcfr) [inline Modules/Loadable/Splines/Logic/vtkSlicerSplinesLogic.cxx:165]: Remove commented code ?

File: Modules/Loadable/Splines/Logic/vtkSlicerSplinesLogic.cxx [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+vtkSlicerSplinesLogic::vtkSlicerSplinesLogic()
+{
+}
+
+//----------------------------------------------------------------------------
+vtkSlicerSplinesLogic::~vtkSlicerSplinesLogic()
+{
+}
+
+void vtkSlicerSplinesLogic::RegisterNodes() {}
```
Comment (jcfr) [inline Modules/Loadable/Splines/Logic/vtkSlicerSplinesLogic.cxx:62]: Remove this and add empty implementation in header ?

**Comment by jcfr** [comment]
Next, I suggest to fix the toggling Polyline vs Spline. 

For reference, the property is available in Slicer here:
image

Relevant code in Slicer:
https://github.com/Slicer/Slicer/blob/f2c326c4d10f2cc2def302caff05faf35db356bd/Modules/Loadable/Markups/MRML/vtkMRMLMarkupsCurveNode.h#L184-L193



Once we will work on revisiting the explore vs annotate to account for the new behavior.

**Comment by allemangD** [comment]
A screenshot below shows the new state of the UI. Note the new "plus" button at the top - this is the new "place" mode.

I included some documentation about the semantics for the new modes in CHANGES.md. I'm unsure where a more suitable place would be - perhaps a new section in README.md?

> - The existing "Annotate" interaction mode now for editing existing control points.
> - The new "Place" interaction mode is for adding new control points.
> - The annotation snaps to the slice in both of these modes.

CellLocator-place-mode-screenshot

----

The spline/polyline toggle is also functioning correctly, and all models update accordingly.

----


With these features, the only remaining bugs are that some of the keyboard shortcuts listed in the readme are still not functioning correctly. Below are the shortcuts that still need fixing:

General:

`Ctrl+W` removes everything from 3d view, leaving it blank.

2D view (Annotation):

These should be available only in the new "Annotate" mode.

`Delete` works only after a point has been moved.
`Ctrl + LMB` does nothing.
`LMB near annotation line` does nothing.
`RMB near annotation line` does nothing.

----

Everything else seems to be at feature parity from my testing. @jcfr do you have any thoughts?

**Comment by jcfr** [comment]
After user is done adding an annotation, mode should automatically go from "The new "Place" interaction" mode to "Annotate" mode


> removes everything from 3d view, leaving it blank.

It seems to work well for me. The CCF atlas is re-loaded as expected.

> perhaps a new section in README.md?

I suggest to add a file called `Documentation/UserGuide.md` 

Shortcuts, description of the modes etc ... would be moved there.

File: Modules/Scripted/Home/Home.py [hunk 4]
```diff
@@ -1151,40 +1159,39 @@ def interactionState(self, newState):
...
-    interactionNode = slicer.app.applicationLogic().GetInteractionNode()
-    if interactionNode.GetCurrentInteractionMode() == interactionNode.Place:
-      return 'annotate'
-    else:
-      return 'explore'
+    return self._interactionState
 
   def setInteractionState(self, newState):
+    newState = newState.lower()
+    if newState not in ('explore', 'annotate', 'place'):
```
Comment (jcfr) [inline Modules/Scripted/Home/Home.py:1166]: What about:

```python
assert newState in ('explore', 'annotate', 'place')
``

**Comment by jcfr** [comment]
Here is the seg fault reported on shutdown:

image

I followed this approach to generate the core dump.
See https://www.slicer.org/wiki/Documentation/Nightly/Developers/Tutorials/Debug_Instructions#Analyze_a_segmentation_fault

**Comment by jcfr** [comment]
Updating Slicer doesn't address the crash

**Comment by jcfr** [comment]
I observed an issue. Here are the steps to reproduce:

1. draw an annotation with few points
2. switch to explore mode and move around
3. go back to annotate

Current behavior: Annotation is snapped to the current location of slice

Expected behavior: slice should be snapped to the location of the annotation so that it can be edited

**Comment by allemangD** [comment]
I attempted to resolve the CHANGES.md conflict by rebasing on the updated master; however GitHub seems to still indicate there are conflicts. For now I'll focus on addressing the other bugs mentioned above instead.

**Pull request merged**